### PR TITLE
New Data Type + Action: Automatic geographic attribute visualizations 

### LINKED ISSUE #123: New Action: Automatic geographic attribute visualizations
For geo-related attributes (longitude, latitude, state, countries, etc), we should detect it as a special data type and display a choropleth automatically.

### PR DESCRIPTION
**Pull request by micahtyong**
## In this PR

Closes #123 by adding a new data type, `geographical`, which detects columns representing geographic attributes (e.g. `state`, `country`), then calls the default action, `univariate`, which then aggregates quantitative attributes and displays the corresponding [choropleth symbol map](https://en.wikipedia.org/wiki/Choropleth_map). 

## Changes
- Addition of new `geographical` datatype
- Modified the `univariate` default action and added `SymbolMap` visualization
- Utilizes existing aggregations from `PandasExecutor.py`
- Tests in `tests/test_type.py`, `tests/test_action.py`, and `test_vegalite_default_actions_registered` for correctness and failure cases

## Example Output
Screen Shot 2021-02-20 at 6 03 58 PM

### CODE REVIEW

**Comment by micahtyong** [comment]
cc: @dorisjlee. Just worked through [this issue](https://github.com/lux-org/lux/issues/123)! Lux can now detect and visualize geographical attributes. I may have pulled upstream changes in a way s.t. this PR includes merged commits from other PRs, but the logic from those remains unchanged. Let me know if there are changes I might need to make. I hope this helps!

File: lux/executor/Executor.py
Comment (dorisjlee) [inline lux/executor/Executor.py:60]: Can we rename the data type name from`geoshape` to `geographical`?

File: lux/executor/Executor.py
Comment (dorisjlee) [inline lux/executor/Executor.py:77]: In the current system, secondary geo attributes like `state` are still declared as nominal. I wonder if we should be labeling them as geo attributes also. If so this may need to account for non-quantitative geo types?
For now, let's take out the geo data type from measures so that they are not plotting in the correlation tab in the example shown.

File: tests/test_type.py
Comment (dorisjlee) [inline tests/test_type.py:104]: `vega_datasets` is not a test dependency. We can just link to the dataset directly: 
```python
df = pd.read_csv('https://raw.githubusercontent.com/altair-viz/vega_datasets/master/vega_datasets/_data/airports.csv')
```

File: lux/action/map.py
Comment (dorisjlee) [inline lux/action/map.py:57]: Would other measure variables be also plotted based on this intent?

**Review by dorisjlee — CHANGES_REQUESTED** [review]

File: tests/test_type.py
Comment (micahtyong) [reply tests/test_type.py:104]: Gotcha, thanks for the tip!

File: lux/action/map.py
Comment (micahtyong) [reply lux/action/map.py:57]: No, since the `geomap` action specifically looks for attributes that are typed `geographical`.

File: lux/executor/Executor.py
Comment (micahtyong) [reply lux/executor/Executor.py]: I see. I may have to take a closer look at this and ask you about it during our 1:1 since geographical attributes being recognized depend on it being part of the measures.

File: tests/test_vis.py
Comment (caitlynachen) [inline tests/test_vis.py:485]: have you looked into Matplotlib supported Geographic charts?

Comment (caitlynachen) [reply tests/test_vis.py:485]: just wondering for future PRs:)

Comment (micahtyong) [reply tests/test_vis.py:485]: Hi @caitlynachen! Yes (: Matplotlib has an API called [basemap](https://matplotlib.org/basemap/) which allows you to make geographic map projections just like the one I did in Altair. There's a good example of the proportional symbol map in [this article](https://jakevdp.github.io/PythonDataScienceHandbook/04.13-geographic-data-with-basemap.html). Let me know what you think!

File: lux/interestingness/interestingness.py
Comment (dorisjlee) [inline lux/interestingness/interestingness.py:369]: What is the goal of the `n_distinct`? I didn't quite understand this and was wondering if we should just assign dummy scores, since it is unlikely that a dataset would have too many geo attributes. Similar to what we do in `generalize.py`.

File: lux/vislib/altair/SymbolMap.py
Comment (dorisjlee) [inline lux/vislib/altair/SymbolMap.py:18]: We should list us and pycountry into the requirement.txt. 
I installed these via pip and noticed that pycountry in particular is quite a big installation (10MB). Perhaps we could just use a dictionary for doing this lookup to keep the package minimal?

File: lux/vislib/altair/SymbolMap.py
Comment (dorisjlee) [inline lux/vislib/altair/SymbolMap.py:25]: Should we rename this GeoHeatmap or Choropleth now that we replaced SymbolMap with the heatmaps?

File: lux/utils/utils.py
Comment (dorisjlee) [inline lux/utils/utils.py:117]: This is a fine implementation for now. Perhaps in the next PR, we can think about making longitude and latitude a geo type also but somehow labelled differently than `state` and `country`. Since that way longitude and latitude would not get lumped with other quantitative types, to be displayed in `Correlation`. 

image

**Review by dorisjlee — CHANGES_REQUESTED** [review]

File: lux/interestingness/interestingness.py
Comment (micahtyong) [reply lux/interestingness/interestingness.py:369]: `n_distinct` computes how many unique values there are for a geographical data type. For example, if a dataset displayed earthquake magnitudes across 48 states and 3 countries, `n_distinct` would return 48 and 3 respectively, meaning the states choropleth would be displayed first, and the countries choropleth second.

Comment (dorisjlee) [reply lux/interestingness/interestingness.py:369]: Ah ok, this makes sense!

File: lux/vislib/altair/SymbolMap.py
Comment (micahtyong) [reply lux/vislib/altair/SymbolMap.py:18]: Sounds good, I've swapped `pycountry` out with [iso3166](https://pypi.org/project/iso3166/#description), which enables me to do the same thing with only 36 kb (much much lighter)

File: lux/vislib/altair/SymbolMap.py
Comment (micahtyong) [reply lux/vislib/altair/SymbolMap.py]: Yup, changed this in the latest commit!

**Comment by micahtyong** [comment]
Hi @dorisjlee, I've added the ability to export choropleths, as requested. An example output is shown below. Let me know if there's anything else I should modify! 

Screen Shot 2021-02-25 at 1 13 50 AM

**Comment by dorisjlee** [comment]
Hi @micahtyong, 
This looks really great! I tried this out on the Covid dataset that we worked on a while back and the maps look really cool with the different measure values in the dataset. 
```python
df = pd.read_csv("https://github.com/covidvis/covid19-vis/blob/master/data/interventionFootprintByState.csv?raw=True",index_col=0)
df['dateBefore'] = pd.to_datetime(df['dateBefore'], format='%Y-%M-%d')
df
```
image
One thing that I did notice from this dataset is that when we click on the visualization, there is a lot of "jitter" that happen with the plots being redrawn. 
Feb-27-2021 21-49-24
I believe this is related to [lux-widget:#63](https://github.com/lux-org/lux-widget/issues/63). This re-drawing issue is present for all charts, but perhaps because the map vis takes longer to draw, so the jitter looks more off. This is a bigger issue that should be addressed separately, so I think we should be good to merge this issue independent of this.

I'll wait until @caitlynachen does a code review on this PR before merging this in. 
Thanks!

**Review by dorisjlee — APPROVED** [review]

File: lux/interestingness/interestingness.py
Comment (caitlynachen) [inline lux/interestingness/interestingness.py:115]: remove change?

**Review by caitlynachen — APPROVED** [review]
Looks great! Maybe next steps is to work on the same geo feature for Matplotlib, but amazing job figuring it out for Altair!

File: lux/vislib/altair/Choropleth.py
Comment (caitlynachen) [inline lux/vislib/altair/Choropleth.py:89]: Is this formatting expected? Make sure it passes "black ."?

**Review by dorisjlee — APPROVED** [review]

**Comment by dorisjlee** [comment]
Thanks @micahtyong!

**Pull request merged**
